File: androidApp/src/main/kotlin/com/malaria/android/ui/onboarding/admin/AdminBiometricStep.kt
package com.malaria.android.ui.onboarding.admin

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Face
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Snackbar
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.fragment.app.FragmentActivity
import com.malaria.android.services.BiometricPrompter
import com.malaria.android.ui.locals.LocalOnboardingState
import com.malaria.android.ui.onboarding.components.WizardStepContainer
import kotlinx.coroutines.launch

/**
 * Step 7 of admin provisioning (spec §10): admin biometric registration.
 *
 * Launches [BiometricPrompter] which wraps `BiometricPrompt` with
 * BIOMETRIC_STRONG | DEVICE_CREDENTIAL — the OS authoritatively confirms a
 * usable biometric or device PIN/passcode is enrolled. On success,
 * `completeAdminProvisioning()` creates the admin `ClinicianProfile`,
 * marks it `biometricEnrolled`, and advances to
 * [com.malaria.android.services.OnboardingState.AdminStep.ProvisioningComplete].
 *
 * Mirrors `iosApp/Views/Onboarding/Admin/AdminBiometricStep.swift`.
 */
@Composable
fun AdminBiometricStep() {
    val onboarding = LocalOnboardingState.current
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val snackbarHostState = remember { SnackbarHostState() }

    var inFlight by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(errorMessage) {
        val message = errorMessage ?: return@LaunchedEffect
        snackbarHostState.showSnackbar(message)
        errorMessage = null
    }

    WizardStepContainer(
        title = "Register administrator biometric",
        subtitle = "This authorizes admin-only actions like changing the threshold or wiping the device. It is not used for routine screening.",
        stepIndicator = "Step 7 of 8",
        primaryLabel = if (inFlight) "Verifying…" else "Register biometric",
        primaryEnabled = !inFlight,
        onPrimary = {
            val activity = context as? FragmentActivity
                ?: run {
                    errorMessage = "Biometric registration unavailable on this activity host."
                    return@WizardStepContainer
                }
            inFlight = true
            scope.launch {
                val prompter = BiometricPrompter(activity)
                val outcome = prompter.prompt(
                    title = "Register administrator biometric",
                    subtitle = "Use your fingerprint, face, or device PIN.",
                )
                when (outcome) {
                    BiometricPrompter.Outcome.Success -> {
                        runCatching { onboarding.completeAdminProvisioning() }
                            .onFailure { errorMessage = it.message ?: "Provisioning failed." }
                    }
                    is BiometricPrompter.Outcome.Failure -> {
                        errorMessage = outcome.reason
                    }
                    BiometricPrompter.Outcome.Cancelled -> {
                        // No-op — let the user try again.
                    }
                }
                inFlight = false
            }
        },
    ) {
        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            androidx.compose.foundation.layout.Row(
                verticalAlignment = Alignment.Top,
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                Icon(Icons.Filled.Face, contentDescription = null)
                Text(
                    text = "Tap the button below. Use the fingerprint sensor, face unlock, or your device PIN when prompted.",
                    style = MaterialTheme.typography.bodyMedium,
                )
            }
            androidx.compose.foundation.layout.Row(
                verticalAlignment = Alignment.Top,
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                Icon(Icons.Filled.Lock, contentDescription = null)
                Text(
                    text = "The biometric is registered with the operating system, not with this app. We never store the fingerprint or face data.",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }

        SnackbarHost(
            hostState = snackbarHostState,
            modifier = Modifier.padding(top = 8.dp),
        ) { data -> Snackbar(snackbarData = data) }
    }
}

File: iosApp/Views/Onboarding/Admin/AdminBiometricStep.swift
import LocalAuthentication
import SwiftUI

/// Step 7 of admin provisioning (spec §10): biometric registration.
///
/// The "registration" happens implicitly: we call
/// `LAContext.evaluatePolicy(.deviceOwnerAuthentication, …)` and a
/// successful evaluation proves the operator has a usable biometric or
/// passcode on the device. Per spec §9 the same `LAContext` policy is
/// reused for routine unlock — no per-app biometric enrolment exists.
///
/// On success we call `OnboardingState.completeAdminProvisioning()` which
/// creates the admin `ClinicianProfile`, marks it biometricEnrolled, and
/// flips the phase to `.microscopistClaim`.
struct AdminBiometricStep: View {

    @Environment(OnboardingState.self) private var onboarding

    @State private var inFlight = false
    @State private var lastError: AuthError?

    var body: some View {
        WizardStepContainer(
            title: "Register administrator biometric",
            subtitle: "This authorizes admin-only actions like changing the threshold or wiping the device. It is not used for routine screening.",
            stepIndicator: "Step 7 of 8",
            primary: .init(title: inFlight ? "Verifying…" : "Register biometric", isEnabled: !inFlight) {
                runBiometric()
            }
        ) {
            VStack(alignment: .leading, spacing: 12) {
                Label("Tap the button below. Use Face ID, Touch ID, or your device passcode when prompted.", systemImage: "faceid")
                    .font(.body)
                Label("The biometric is registered with the operating system, not with this app. We never store the fingerprint or face data.", systemImage: "lock.shield")
                    .font(.body)
                    .foregroundStyle(.secondary)
            }
        }
        .alert(
            "Biometric registration failed",
            isPresented: Binding(
                get: { lastError != nil },
                set: { if !$0 { lastError = nil } }
            )
        ) {
            Button("OK", role: .cancel) { lastError = nil }
        } message: {
            Text(lastError?.errorDescription ?? "")
        }
    }

    private func runBiometric() {
        inFlight = true
        Task { @MainActor in
            defer { inFlight = false }
            let laContext = LAContext()
            var error: NSError?
            guard laContext.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error) else {
                lastError = .noBiometricsConfigured
                return
            }
            do {
                try await laContext.evaluatePolicy(
                    .deviceOwnerAuthentication,
                    localizedReason: "Register administrator biometric"
                )
            } catch let laError as LAError {
                switch laError.code {
                case .userCancel, .systemCancel, .appCancel:
                    lastError = .userCancelled
                default:
                    lastError = .authenticationFailed(reason: laError.localizedDescription)
                }
                return
            } catch {
                lastError = .authenticationFailed(reason: error.localizedDescription)
                return
            }

            do {
                _ = try onboarding.completeAdminProvisioning()
            } catch {
                lastError = .authenticationFailed(reason: error.localizedDescription)
            }
        }
    }
}
